src/test/scala/is/hail/distributedmatrix/BlockMatrixSuite.scala

```diff
@@ -0,0 +1,545 @@
+    val n = rows.length
+    val m = if (n == 0) 0 else rows(0).length
+
+    BlockMatrix.from(sc, new BDM[Double](m, n, rows.flatten.toArray).t, blockSize)
+  }
+
+  def toBM(lm: BDM[Double]): BlockMatrix =
+    toBM(lm, BlockMatrix.defaultBlockSize)
+
+  def toBM(lm: BDM[Double], blockSize: Int): BlockMatrix =
+    BlockMatrix.from(sc, lm, blockSize)
+
+  def blockMatrixPreGen(blockSize: Int): Gen[BlockMatrix] =
+    Gen.coin().flatMap(blockMatrixPreGen(blockSize, _))
+
+  def blockMatrixPreGen(blockSize: Int, transposed: Boolean): Gen[BlockMatrix] = for {
+    (l, w) <- Gen.nonEmptySquareOfAreaAtMostSize
+    m <- blockMatrixPreGen(l, w, blockSize, transposed)
+  } yield m
+
+  def blockMatrixPreGen(rows: Int, columns: Int, blockSize: Int): Gen[BlockMatrix] =
+    Gen.coin().flatMap(blockMatrixPreGen(rows, columns, blockSize, _))
+
+  def blockMatrixPreGen(rows: Int, columns: Int, blockSize: Int, transposed: Boolean): Gen[BlockMatrix] = for {
+    arrays <- Gen.buildableOfN[Seq, Array[Double]](rows, Gen.buildableOfN(columns, arbDouble.arbitrary))
+    m = toBM(arrays, blockSize)
+  } yield if (transposed) m.t else m
+
+  val squareBlockMatrixGen = for {
+    size <- Gen.size
+    l <- Gen.interestingPosInt
+    s = math.sqrt(math.min(l, size)).toInt
+    blockSize <- Gen.interestingPosInt
+    m <- blockMatrixPreGen(s, s, blockSize)
+    _ = println(s"$l $s $blockSize")
+  } yield m
+
+  val blockMatrixGen = for {
+    blockSize <- Gen.interestingPosInt
+    m <- blockMatrixPreGen(blockSize)
+  } yield m
+
+  val twoMultipliableBlockMatrices = for {
+    Array(rows, inner, columns) <- Gen.nonEmptyNCubeOfVolumeAtMostSize(3)
+    blockSize <- Gen.interestingPosInt
+    l <- blockMatrixPreGen(rows, inner, blockSize)
+    r <- blockMatrixPreGen(inner, columns, blockSize)
+  } yield (l, r)
+
+  implicit val arbitraryHailBlockMatrix =
+    Arbitrary(blockMatrixGen)
+
+  @Test
+  def pointwiseSubtractCorrect() {
+    val m = toBM(4, 4, Array[Double](
+      1,  2,  3,  4,
+      5,  6,  7,  8,
+      9,  10, 11, 12,
+      13, 14, 15, 16))
+
+    val expected = toLM(4, 4, Array[Double](
+      0, -3, -6, -9,
+      3, 0,  -3, -6,
+      6, 3,  0,  -3,
+      9, 6,  3,  0))
+
+    val actual = (m :- m.t).toLocalMatrix()
+    assert(actual == expected)
+  }
+
+  @Test
+  def multiplyByLocalMatrix() {
+    val ll = toLM(4, 4, Array[Double](
+      1,  2,  3,  4,
+      5,  6,  7,  8,
+      9,  10, 11, 12,
+      13, 14, 15, 16))
+    val l = toBM(ll)
+
+    val lr = toLM(4, 1, Array[Double](
+      1,
+      2,
+      3,
+      4))
+
+    assert(ll * lr === (l * lr).toLocalMatrix())
+  }
+
+  @Test
+  def randomMultiplyByLocalMatrix() {
+    forAll(Gen.twoMultipliableDenseMatrices[Double]) { case (ll, lr) =>
+      val l = toBM(ll)
+      assert(ll * lr === (l * lr).toLocalMatrix())
+      true
+    }.check()
+  }
+
+  private def arrayEqualNaNEqualsNaN(x: Array[Double], y: Array[Double], absoluteTolerance: Double = 1e-15): Boolean = {
```

Comment:
@jbloom22 I honestly feel like this is an ugly function with an ugly name. I'm rather embarrassed to share it with others.

I want some system for defining equivalence that doesn't require me to have a new equivalence function for any type wherein the nested Double's should be treated equal if both are NaN.

I have some ideas for such a system but no time to try them out. :/

Reply:
Ok, let's keep it private for now then